africanus/averaging/baseline_time_and_channel_avg.py

```diff
@@ -0,0 +1,575 @@
+    time_centroid_normaliser = normaliser_factory(have_time_centroid)
+    weight_normaliser = normaliser_factory(have_weight)
+    sigma_normaliser = normaliser_factory(have_sigma)
+    
+    def impl(meta, ant1, ant2, interval=None, flag_row=None, time_centroid=None, 
+            exposure=None, uvw=None, weight=None, sigma=None):
+        
+        out_rows = meta.time.shape[0]
+        print(out_rows)
+        print(uvw.shape)
+        counts = np.zeros(out_rows, dtype=np.uint32)
+        ant1_avg = np.empty(out_rows, ant1.dtype)
+        ant2_avg = np.empty(out_rows, ant2.dtype)
+        
+        interval_avg = interval_factory(out_rows, interval)
+        uvw_avg = uvw_factory(out_rows, uvw)
+        time_centroid_avg = time_centroid_factory(out_rows, time_centroid)
+        exposure_avg = exposure_factory(out_rows, exposure)
+        weight_avg = weight_factory(out_rows, weight)
+        sigma_avg = sigma_factory(out_rows, sigma)
+        
+        # 
+        for in_row, out_row in enumerate(meta.map):
+            if flags_match(flag_row, in_row, meta.flag_row, out_row):
+                interval_adder(interval_avg, out_row, interval, in_row)
+                uvw_adder(uvw_avg, out_row, uvw, in_row)
+                weight_adder(weight_avg, out_row, weight, in_row)
+                sigma_adder(sigma_avg, out_row, sigma, in_row)
+                time_centroid_adder(time_centroid_avg, out_row, time_centroid, in_row)
+                exposure_adder(exposure_avg, out_row, exposure, in_row)
+                
+                counts[out_row] += 1
+            
+            ant1_avg[out_row] = ant1[in_row]
+            ant2_avg[out_row] = ant2[in_row]
+        
+        
+        # Normalise
+        for out_row in range(out_rows):
+            count = counts[out_row]
+            
+            if count > 0:
+                uvw_normaliser(uvw_avg, out_row, count)
+                weight_normaliser(weight_avg, out_row, count)
+                sigma_normaliser(sigma_avg, out_row, count)
+                time_centroid_normaliser(time_centroid_avg, out_row, count)
+        
+        return RowAverageOutput(interval_avg, ant1_avg, ant2_avg,
+                                time_centroid_avg, exposure_avg, uvw_avg,
+                                weight_avg, sigma_avg)
+    return impl
+
+
+
+
+def vis_ampl_output_factory(present):
+    """ Returns function producing output vis amplitudes if present """
+    if present:
+        def impl(shape, array):
+            return np.zeros(shape, dtype=array.real.dtype)
+    else:
+        def impl(shape, array):
+            pass
+
+    return njit(nogil=True, cache=True)(impl)
+
+
+def chan_output_factory(present):
+    """ Returns function producing outputs if the array is present """
+    if present:
+        def impl(shape, array):
+            return np.zeros(shape, dtype=array.dtype)
+    else:
+        def impl(shape, array):
+            pass
+
+    return njit(nogil=True, cache=True)(impl)
```

Comment:
@sjperkins can you help me look into this. I get numba compilation error on the 'chan_output_factory' function

```
Traceback (most recent call last):
  File "test_baseline_time_and_channel.py", line 76, in <module>
    bins_for_longest_baseline=1.0, baseline_chan_bin_size=1)
  File "/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py", line 376, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py", line 343, in error_rewrite
    reraise(type(e), e, None)
  File "/anaconda3/lib/python3.7/site-packages/numba/six.py", line 658, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<built-in function zeros>) with argument(s) of type(s): ((int64, array(uint32, 1d, C), int64), dtype=dtype(complex64))
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<built-in function zeros>)
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (194)


File "../baseline_time_and_channel_avg.py", line 194:
        def impl(shape, array):
            return np.zeros(shape, dtype=array.dtype)
            ^

[1] During: resolving callee type: type(CPUDispatcher(<function chan_output_factory.<locals>.impl at 0x619b0abf8>))
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (348)

[3] During: resolving callee type: type(CPUDispatcher(<function chan_output_factory.<locals>.impl at 0x619b0abf8>))
[4] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (353)

[5] During: resolving callee type: type(CPUDispatcher(<function chan_output_factory.<locals>.impl at 0x619b0abf8>))
[6] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (348)


File "../baseline_time_and_channel_avg.py", line 348:
    def impl(row_meta, chan_meta,flag_row=None, vis=None, flag=None,
        <source elided>

        vis_avg = vis_factory(out_shape, vis)
        ^

[1] During: resolving callee type: type(CPUDispatcher(<function baseline_chan_average at 0x11927d730>))
[2] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (558)

[3] During: resolving callee type: type(CPUDispatcher(<function baseline_chan_average at 0x11927d730>))
[4] During: typing of call at /Users/smasoka/Varsity/codex-africanus/africanus/averaging/baseline_time_and_channel_avg.py (558)


File "../baseline_time_and_channel_avg.py", line 558:
    def impl(time, interval, antenna1, antenna2,
        <source elided>
                                          flag=flag, weight_spectrum=weight_spectrum, sigma_spectrum=sigma_spectrum,
                                          baseline_chan_bin_size=baseline_chan_bin_size)
                                          ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/latest/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

Reply:
I tried to reproduce to try understand but

```
In [2]: import numpy as np 
   ...: from numba import njit 
   ...: from numba import generated_jit 
   ...:  
   ...: nchan = 16 
   ...: ncorr = 4 
   ...: vis = (np.arange(10*nchan*ncorr, dtype=np.float32) + 
   ...:       np.arange(1, 10*nchan*ncorr+1, dtype=np.float32)*1j) 
   ...: vis = vis.reshape(10, nchan, ncorr) 
   ...:  
   ...: shape = (10, nchan, ncorr) 
   ...: np.zeros(shape, dtype=vis.dtype) 
   ...:  
   ...: @generated_jit(nopython=True, nogil=True, cache=True) 
   ...: def call_impl(shape, vis): 
   ...:     impl(shape, vis) 
   ...:  
   ...: @njit(nogil=True, cache=True) 
   ...: def impl(shape, array): 
   ...:     return np.zeros(shape, dtype=array.dtype) 
   ...:                                                                                                                                                       

In [3]: call_impl(shape,vis)                                                                                                                                  
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-3-25a1767fcd6c> in <module>
----> 1 call_impl(shape,vis)

/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    374                 e.patch_message(msg)
    375 
--> 376             error_rewrite(e, 'typing')
    377         except errors.UnsupportedError as e:
    378             # Something unsupported is present in the user code, add help info

/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    341                 raise e
    342             else:
--> 343                 reraise(type(e), e, None)
    344 
    345         argtypes = []

/anaconda3/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'dtype' of type typeref[array(complex64, 3d, C)]

File "<ipython-input-2-f6e53a862b5e>", line 20:
def impl(shape, array):
    return np.zeros(shape, dtype=array.dtype)
    ^

[1] During: typing of get attribute at <ipython-input-2-f6e53a862b5e> (20)

File "<ipython-input-2-f6e53a862b5e>", line 20:
def impl(shape, array):
    return np.zeros(shape, dtype=array.dtype)
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.
.......
```